sklearn/linear_model/ridge.py

```diff
@@ -141,7 +145,17 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
         # Only one penalty, we can solve multi-target problems in one time.
         K.flat[::n_samples + 1] += alpha[0]
 
-        dual_coef = linalg.solve(K, y, sym_pos=True, overwrite_a=True)
+        try:
+            # Note: we must use overwrite_a=False in order to be able to
+            #       use the fall-back solution below in case a LinAlgError
+            #       is raised
+            dual_coef = linalg.solve(K, y, sym_pos=True,
+                                     overwrite_a=False)
+        except np.linalg.LinAlgError:
+            warnings.warn("Singular matrix in solving dual problem. Using "
+                          "pseudo inverse instead.")
+            Kinv = linalg.pinv2(K)
+            dual_coef = np.dot(Kinv, y)
```

Comment:
wrt my comment: calling `pinv2` or `lstsq` basically boils down to doing an svd on `K + ridge`. So from that point of view it should be the one or the other. However, looking at the underlying lapack functions used, `lstsq` calls `gelss` ("solve least squares using SVD") and `pinv2` calls scipy svd which in turn calls `gesdd` ("do svd using a divide and conquer method"). What I can't discern is whether maybe `lstsq` manages to avoid some calculation, and also on what svd implementation it is based. In any case: I don't think it makes sense to give the option to decide between `pinv2` and `lstsq` and in doubt it should be the function dedicated to solving this type of problem, i.e. `lstsq`.

As far as other solvers goes, yes, there are other ways of solving kernel ridge, e.g. by re-using an eigendecomposition of `K` for potentially different penalties to name only one. But shouldn't adding more solvers and decision logic should maybe be kept for after merging this? (I think this code already provides enough functionality to be considered a valid contribution.)